README.md
# Cap height

Calculate the **cap height** of any font loaded with [Web Font Loader](https://github.com/typekit/webfontloader). See the demo on [CodePen](http://codepen.io/sebdesign/pen/EKmbGL?editors=0011).

**What is the cap height?**

The cap height is the height of a capital letter measured from the baseline.

> Read more on [Typography Deconstructed](http://www.typographydeconstructed.com/cap-height) and on [Wikipedia](https://en.wikipedia.org/wiki/Cap_height).

**How is this useful?**

The web lacks *exquisite* typography. Of course, its state is very good with the wide use of `@font-face` and the quality web font distribution services and foundries. However, setting type on the web still challenging. [Several](https://github.com/StudioThick/megatype) [attempts](http://compass-style.org/reference/compass/typography/vertical_rhythm) [and](https://gist.github.com/razwan/10662500) [tools](https://sassline.com) have risen in order to define a proper vertical rhythm and a baseline grid.

But most of them depend on a *cap height* ratio, which varies for every typeface and cannot be calculated with CSS. You have to rely on design software or font metrics from the foundries to find this out. This package addresses this issue by rendering loaded fonts on a canvas to measure the actual height of the glyphs, and thus calculating the cap-height/font-size ratio. Then you could easily apply this value to any tool you like.

**Read more about the subject**

- [Web typography is broken. Here’s how we can fix it](http://www.studiothick.com/essays/web-typography-is-broken)

- [Aligning type to baseline the right way using SASS](https://medium.com/written-in-code/aligning-type-to-baseline-the-right-way-using-sass-e258fce47a9b)

## Demo

See the demo on [CodePen](http://codepen.io/sebdesign/pen/EKmbGL?editors=0011). Open the CodePen console to see the output.

Or run `dist/index.html` in your browser and open the developer tools. In the output of the console you should see a JSON object for each loaded font, with its properties. An additional `--cap-height` CSS property contains the **holy grail** of the web typography.

In the browser window a canvas will be displayed for each loaded font with the letter *H*.

This example loads *Roboto Mono* through *Google Fonts*; but you can use any module available in the *Web Font Loader*, like *Typekit*, *Fondeck*, *Fonts.com* or your own web fonts.

> Check all the available modules: https://github.com/typekit/webfontloader#modules

## How it works

When a `@font-face` font is loaded with the *Web Font Loader*, it creates a HTML canvas and draws the letter *H* with the appropriate CSS properties extracted from the [Font Variation Description](http://typekit.github.io/fvd). Then it measures the actual pixel height of the drawn letter, and calculates the ratio between the height and the `font-size`.

## WIP

This project is a work in progress. I’m planning to add more methods to calculate loaded fonts beyond the *Web Font Loader*, like the [Font Face API](https://developer.mozilla.org/el/docs/Web/API/FontFace), or simply by selecting a DOM element along with its CSS properties.

## Installation

### In a browser:

Include the `dist/cap-height.js` script in your document, then call the methods on the `capHeight` instance. **Don’t forget to include the Web Font Loader**.

```html
<script src="https://unpkg.com/lodash@4.17.4"></script>
<script src="./cap-height.js"></script>

<script>
  capHeight.calculate({});
</script>
```

### As a module:

```shell
$ npm install cap-height
```

```js
import capHeight from "capHeight";

capHeight.calculate({});
```

## API

### calculate(properties, [text=H])

This is the main method which calculates the `cap-height`. The first argument is an object containing the `font-style`, `font-weight`, `font-size`, and `font-family` properties. If you omit any of these, the defaults in the example below will be used.

**Warning**: The font-size **MUST** be in pixels. If any other unit of unit-less number is passed, it will be evaluated in pixels. E.g. `2em` or `2` will be computed to `2px`.

The second argument is optional. It defines the text that will be used to measure the cap-height. The default is *H*.

The method returns the passed object along with a `--cap-height` CSS property, and any other omitted property.

```js
capHeight.calculate({
  "font-style": "normal",
  "font-weight": 400,
  "font-size": "100px",
  "font-family": "serif"
}, "HI");

// Output
{
  "font-style": "normal",
  "font-weight": 400,
  "font-size": "100px",
  "font-family": "serif",
  "--cap-height": 0.66
}
```

### fontActive(callback, text)

Event handler for the `fontactive` or `active` events of the *Web Font Loader*.

This method accepts a `callback` function, in which the calculated properties will be passed. The callback is executed after the Web Font Loader has fired the appropriate events and the calculations are completed.

> Read about the Web Font Loader events: https://github.com/typekit/webfontloader#events

You can pass a string as an optional second argument after your callback, to define the text that will be used to calculate the cap-height. The default is *H*.

```js
WebFontConfig = {
  {
    // define your font modules here
  },
  fontactive: capHeight.fontActive(function(properties) {
    console.log(properties);
  }, "HI");
};
```

### setContainer(element)

If you want to see/debug the process, define a DOM element in which a canvas(es) of each rendered font will be displayed.

```js
WebFontConfig = {
  // set the container while the fonts are loading
  loading: function() {
    capHeight.setContainer(document.body);
  }
}
```

## Credits

- [Thomas Bredin-Grey](https://github.com/tbredin) for the essay [Web typography is broken. Here’s how we can fix it](http://www.studiothick.com/essays/web-typography-is-broken/), which was the main inspiration for this project.

- [Sebastian Lasse](https://github.com/sebilasse) for the [font metrics for google fonts](http://codepen.io/sebilasse/pen/gPBQqm?editors=1010) on CodePen, which provided the main implementation the cap-height calculation.

- [Percolate](https://github.com/percolate) for the [Font Variation Description for JavaScript](https://github.com/percolate/fvd) module.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## [2.0.0] - 2017-02-19
The `cap-height` property has been renamed to `--cap-height` to match custom CSS properties.

The codebase has been refactored using a more functional style and ES6.

Migrated from Browserify to webpack 2.

Added MIT license and changelog.

## [1.1.2] - 2016-03-24
Support high DPI rendering on canvas .

## [1.1.1] - 2016-03-23
Added a demo on [CodePen](http://codepen.io/sebdesign/pen/EKmbGL?editors=0011).

## [1.1.0] - 2016-03-23
Refactoring and docs updated.

## [1.0.0] - 2016-03-22
Initial release.


lib/capHeight.js
// var WebFont = require("webfontloader");
var _ = require("lodash");
var _fvd = require("fvd");
var format = require("string-format");

/**
 * The threshold to determine non-white pixels.
 *
 * @type Number
 */
var threshold = 0.75;

/**
 * The device pixel ratio, used to calculate the dimensions on high-DPI screens.
 *
 * @type Number
 */
var dpr = window.devicePixelRatio;

/**
 * The element that will contain the canvases.
 *
 * @type Node
 */
var container;

module.exports = {
  calculate: calculate,

  /**
   * Event listener for the fontactive event of the Web Font Loader.
   *
   * Calculate the cap-height and execute the given callback with the resulted properties.
   *
   * @param  Function callback
   * @param  String   text
   * @return Function
   */
  fontActive: function(callback, text) {
    return function(familyName, fvd) {
      // Parse the fvd properties returned from the Web Font Loader
      var parsed = _fvd.parse(fvd);

      // Add the font-family
      _.set(parsed, "font-family", familyName);

      // Calculate the cap-height
      var properties = calculate(parsed, text);

      // Execute the callback
      return callback(properties);
    };
  },

  /**
   * Set a container element to display the rendered canvas.
   *
   * @param  Node element
   * @return Node
   */
  setContainer: function(element) {
    return container = element;
  }
};

/**
 * Calculate the cap height.
 *
 * @param Object properties
 * @param String text
 * @return Object
 */
function calculate(properties, text) {
  // Merge with the default properties
  properties = mergeWithDefaults(properties);

  // Measure the actual size of the text
  var height = measureTextHeight(properties, text);

  // Calculate the ratio between the height and the font-size
  var capHeight = height / getFontSize(properties);

  // Add the cap-height to the properties
  _.set(properties, "cap-height", capHeight);

  return properties;
}

/**
 * Merge the given properties with the defaults.
 *
 * @param  Object properties
 * @return Object
 */
function mergeWithDefaults(properties) {
  var defaults = {
    "font-style": "normal",
    "font-weight": 400,
    "font-size": "100px",
    "font-family": "serif"
  };

  return _.merge(defaults, properties);
}

/**
 * Measure the height of the given text with the specified font properties.
 *
 * @param  Object properties
 * @param  String text
 * @return Number
 *
 * @throws RangeError
 */
function measureTextHeight(properties, text) {
  var fontSize = getFontSize(properties);

  text = text || "H";

  if (/\s/.test(text)) {
    throw new RangeError("Cannot calculate the height of whitespace.");
  }

  var base = 2 * fontSize;

  // Set the actual pixel dimensions
  var dimensions = {
    width: base * text.length,
    height: base
  };

  // Set the physical pixel dimensions
  var physicalDimensions = _.mapValues(dimensions, toDot);

  var canvas = createCanvas(dimensions, physicalDimensions);
  var context = getContext(canvas);

  context = drawContext(context, properties, dimensions, text);

  var pixelData = getPixelData(context, physicalDimensions);
  var ascent = getAscent(pixelData, physicalDimensions.width);
  var descent = getDescent(pixelData, physicalDimensions.width);
  var capHeight = getHeight(descent, ascent);

  // Draw a rectangle marking the cap height bounds
  context.fillStyle = "rgba(255, 85, 51, .1)";
  context.fillRect(0, toPx(ascent), toPx(physicalDimensions.width), toPx(capHeight));

  // Display the canvas
  displayCanvas(canvas);

  return toPx(capHeight);
}

/**
 * Get the font-size as a number.
 *
 * @param Object properties
 * @return Number
 */
function getFontSize(properties) {
  var fontSize = parseFloat(_.replace(_.get(properties, "font-size", 0), /\D/g, ''));

  if (isNaN(fontSize)) {
    throw new RangeError("Cannot calculate the height of the text with invalid font-size.");
  }

  if (fontSize === 0) {
    throw new RangeError("Cannot calculate the height of the text with font-size: 0.");
  }

  return fontSize;
}

/**
 * Create a canvas with the given dimensions.
 *
 * @param Object dimensions
 * @param Object physicalDimensions
 * @return HTMLCanvasElement
 */
function createCanvas(dimensions, physicalDimensions) {
  var canvas = document.createElement("canvas");

  canvas.width = physicalDimensions.width;
  canvas.height = physicalDimensions.height;

  canvas.style.width = dimensions.width + "px";
  canvas.style.height = dimensions.height + "px";

  canvas.style.margin = "10px";

  return canvas;
}

/**
 * Append the given canvas to the container element.
 *
 * @param HTMLCanvasElement canvas
 * @return void
 */
function displayCanvas(canvas) {
  if (container && container.nodeType) {
    container.appendChild(canvas);
  }
}

/**
 * Get the canvas drawing context.
 *
 * @return CanvasRenderingContext2D
 */
function getContext(canvas) {
  var context = canvas.getContext("2d");

  context.scale(dpr, dpr);

  return context;
}

/**
 * Draw the text and the background in the canvas context.
 *
 * @param CanvasRenderingContext2D context
 * @param Object properties
 * @param Object dimensions
 * @param String context
 * @return CanvasRenderingContext2D
 */
function drawContext(context, properties, dimensions, text) {
  context.font = composeFontProperty(properties);

  // Align the text horizontally and vertically
  context.textAlign = "center";
  context.textBaseline = "middle";

  // Set all canvas pixeldata values to 255, with all the content
  // data being 0. This lets us scan for data[i] != 255.
  context.fillStyle = "white";
  context.fillRect(0, 0, dimensions.width, dimensions.height);
  context.fillStyle = "black";

  // Draw the text at the center
  context.fillText(text, dimensions.width / 2, dimensions.height / 2);

  return context;
}

/**
 * Compose the font shorthand property.
 *
 * @param Object properties
 * @return String
 *
 * @see https://developer.mozilla.org/en-US/docs/Web/CSS/font
 */
function composeFontProperty(properties) {
  return format("{font-style} {font-weight} {font-size} {font-family}", properties);
}

/**
 * Get the pixel data of the context.
 *
 * @param  CanvasRenderingContext2D context
 * @param  Object dimensions
 * @return Uint8ClampedArray
 */
function getPixelData(context, dimensions) {
  return context.getImageData(0, 0, dimensions.width, dimensions.height).data;
}

/**
 * Calculate the height of the text with the given ascent and descent.
 *
 * @param  Number ascent
 * @param  Number descent
 * @return Number
 */
function getHeight(descent, ascent) {
  return descent + 1 - ascent;
}

/**
 * Get the ascent of the text.
 *
 * Forward scan the pixel data and get the vertical position of the first non-white pixel.
 *
 * @param  Uint8ClampedArray pixelData
 * @param  Number width
 * @return Number
 */
function getAscent(pixelData, width) {
  var i = _.findIndex(pixelData, isNotWhite);

  return getYPosition(i, width);
}

/**
 * Get the descent of the text.
 *
 * Reverse scan the pixel data and get the vertical position of the last non-white pixel.
 *
 * @param  Uint8ClampedArray pixelData
 * @param  Number width
 * @return Number
 */
function getDescent(pixelData, width) {
  var i = _.findLastIndex(pixelData, isNotWhite);

  return getYPosition(i, width);
}

/**
 * Get the vertical position of the pixel data index, given the width.
 *
 * @param Number i
 * @param Number width
 * @return Number
 */
function getYPosition(i, width) {
  return _.floor((i / 4) / width);
}

/**
 * Check the given RGBA value is below the threshold.
 *
 * @param Number value
 * @return Boolean
 */
function isNotWhite(value) {
  return value < (255 * threshold);
}

/**
 * Convert a physical pixel value to a CSS pixel
 */
function toPx(value) {
  return value / dpr;
}

/**
 * Convert a CSS pixel to a physical pixel
 */
function toDot(value) {
  return value * dpr;
}


src/index.js
/**
 * Cap height Module.
 *
 * @module cap-height
 */

// import WebFont from 'webfontloader';
import _ from 'lodash';
import _fvd from 'fvd';

/**
 * The threshold to determine non-white pixels.
 *
 * @type {number}
 */
const THRESHOLD = 0.75;

/**
 * The margin to apply between canvases.
 *
 * @type {number}
 */
const CANVAS_MARGIN = 10;

/**
 * Canvas size / font-size ratio.
 *
 * @type {Number}
 */
const MULTIPLIER = 2;

/**
 * The device pixel ratio, used to calculate the dimensions on high-DPI screens.
 *
 * @type {number}
 */
const DPR = window.devicePixelRatio;

/**
 * The element that will contain the canvases.
 *
 * @type {Node}
 */
let container;

/**
 * The default font properties.
 *
 * @type {object}
 */
const defaults = {
  'font-style': 'normal',
  'font-weight': 400,
  'font-size': '100px',
  'font-family': 'serif'
};

_.mixin({
  fillRect(context, x, y, width, height) {
    return _.tap(context, _.method('fillRect', x, y, width, height));
  },
  fillText(context, text, x, y) {
    return _.tap(context, _.method('fillText', text, x, y));
  },
  scale(context, x, y = x) {
    return _.tap(context, _.method('scale', x, y));
  },
});

/**
 * Calculate the cap height.
 *
 * @param  {object} properties The font properties
 * @param  {string} [text=H] The text to apply the properties
 * @return {object} The properties with the cap-height key value
 */
const calculate = (properties, text = 'H') => {
  return _(properties)
    .defaults(defaults)
    .thru(_.partial(measureTextHeight, text))
    .value();
};

module.exports = {
  /**
   * Calculate the cap height.
   */
  calculate,

  /**
   * Calculate the cap-height and execute the given callback with the resulted properties.
   *
   * @param  {function} callback The callback to execute when the calculation is complete.
   * @param  {string}   text The text used to calculate the cap-height.
   * @return {function} The listener for the webfontloader event.
   */
  fontActive(callback, text) {
    /**
     * Listener for the fontactive event of the Web Font Loader.
     *
     * @param {string} familyName The font-family to be added to the properties.
     * @param {object} fvd The parsed font properties.
     * @return {object} The wrapped font properties.
     * @listens module:webfont~fontactive
     */
    return (familyName, fvd) => {
      return _(fvd)
        // Parse the fvd properties returned from the Web Font Loader
        .thru(_fvd.parse)
        // Set the font-family
        .set('font-family', familyName)
        // Execute the calculation
        .thru(_.partialRight(calculate, text))
        // Execute the callback
        .thru(callback)
        .commit();
    };
  },

  /**
   * Inspect a DOM element and calculate the cap-height given its computed CSS properties.
   *
   * @param  {Node}   element The node element to inspect.
   * @param  {string} text The used to calculate the cap-height.
   * @return {object} The font properties.
   */
  inspect(element, text) {
    return _(element)
      .thru(window.getComputedStyle)
      .pick(_.keys(defaults))
      .thru(_.partialRight(calculate, text))
      .value();
  },

  /**
   * Set a container element to display the rendered canvases.
   *
   * @param {Node} element The node element to contain the canvases.
   */
  setContainer(element) {
    container = element;
  }
};

/**
 * Measure the height of the given text with the specified font properties.
 *
 * @param  {string} [text=H] The text being measured.
 * @param  {object} properties The font properties used to measure the height.
 * @return {object} The font properties with the cap-height.
 * @throws {RangeError}
 */
const measureTextHeight = (text = 'H', properties) => {
  const fontSize = getFontSize(properties);
  const base = MULTIPLIER * fontSize;

  if (/\s/.test(text)) {
    throw new RangeError('Cannot calculate the height of whitespace.');
  }

  // Set the actual pixel dimensions
  const dimensions = {
    width: base * text.length,
    height: base
  };

  // Set the physical pixel dimensions
  const physicalDimensions = _.mapValues(dimensions, toDot);

  const canvas = createCanvas(dimensions, physicalDimensions);
  let context = getContext(canvas);

  context = drawContext(context, properties, dimensions, text);

  const metrics = getMetrics(context, physicalDimensions);
  const capHeight = getHeight(metrics);

  // Draw a rectangle marking the cap height bounds
  _(context)
    .set('fillStyle', 'rgba(255, 85, 51, .1)')
    .fillRect(0, toPx(metrics.ascent), toPx(physicalDimensions.width), toPx(capHeight))
    .commit();

  // Display the canvas
  displayCanvas(canvas);

  return _.assign(properties, {
    '--cap-height': toPx(capHeight) / fontSize
  });
};

/**
 * Get the font-size as a unitless number.
 *
 * @param  {object} properties The font properties containing the font-size.
 * @return {number} The font-size as a unitless number.
 * @throws {RangeError}
 */
const getFontSize = (properties) => {
  return _
    .chain(properties)
    .get('font-size', 0)
    .replace(/\D/g, '')
    .toNumber()
    .thru((fontSize) => {
      if (isNaN(fontSize)) {
        throw new RangeError('Cannot calculate the height of the text with invalid font-size.');
      }

      if (fontSize === 0) {
        throw new RangeError('Cannot calculate the height of the text with font-size: 0.');
      }

      return fontSize;
    });
};

/**
 * Create a canvas with the given dimensions.
 *
 * @param  {object} dimensions The width and height to set as CSS styles.
 * @param  {object} physicalDimensions The physical width and height to set to the canvas element.
 * @return {HTMLCanvasElement} The canvas element.
 */
const createCanvas = (dimensions, physicalDimensions) => {
  let style = _(dimensions)
    .set('margin', CANVAS_MARGIN)
    .mapValues(value => value + 'px')
    .value();

  return _(document.createElement('canvas'))
    .assign(physicalDimensions)
    .merge({'style': style})
    .value();
};

/**
 * Append the given canvas to the container element.
 *
 * @param {HTMLCanvasElement} canvas The canvas to add to the container.
 * @return void
 */
const displayCanvas = (canvas) => {
  if (_.isElement(container)) {
    container.appendChild(canvas);
  }
};

/**
 * Get the scaled canvas drawing context.
 *
 * @param  {HTMLCanvasElement} canvas The canvas element.
 * @return {CanvasRenderingContext2D} The canvas context.
 */
const getContext = (canvas) => {
  return _(canvas.getContext('2d'))
    .scale(DPR)
    .value();
};

/**
 * Draw the text and the background in the canvas context.
 *
 * @param {CanvasRenderingContext2D} context The canvas context.
 * @param {object} properties The font properties.
 * @param {object} dimensions The dimensions of the drawing.
 * @param {number} dimensions.width The width of the drawing.
 * @param {number} dimensions.height The height of the drawing.
 * @param {string} text The text to draw.
 * @return {CanvasRenderingContext2D} The canvas context with the rendered text and background.
 */
const drawContext = (context, properties, {width, height}, text) => {
  return _(context)
    // Set all canvas pixeldata values to 255, with all the content
    // data being 0. This lets us scan for data[i] != 255.
    .set('fillStyle', 'white')
    .fillRect(0, 0, width, height)
    // Set the text properties
    .assign({
      font: composeFontProperty(properties),
      fillStyle: 'black',
      textAlign: 'center',
      textBaseline: 'middle'
    })
    // Draw the text at the center
    .fillText(text, width / 2, height / 2)
    .value();
};

/**
 * Compose the font shorthand property.
 *
 * @param  {object} properties The font properties.
 * @return {string} The shorthand font value.
 *
 * @see https://developer.mozilla.org/en-US/docs/Web/CSS/font
 */
const composeFontProperty = (properties) => {
  return _(properties)
    .at('font-style', 'font-weight', 'font-size', 'font-family')
    .join(' ');
};

/**
 * Get the ascent and descent from the pixel data.
 *
 * @param  {CanvasRenderingContext2D} context The canvas context.
 * @param  {object} dimensions The physical dimensions used as bounds.
 * @param  {number} dimensions.width The physical width.
 * @param  {number} dimensions.height The physical height.
 * @return {object} The ascent and descent.
 */
const getMetrics = (context, {width, height}) => {
  let pixelData = getPixelData(context, width, height);

  return {
    ascent: getAscent(pixelData, width),
    descent: getDescent(pixelData, width),
  };
};

/**
 * Get the pixel data of the context.
 *
 * @param  {CanvasRenderingContext2D} context The canvas context.
 * @param  {number} width The width of the extracted area.
 * @param  {number} height The height of the extracted area.
 * @return {Uint8ClampedArray} One-dimensional array containing RGBA data.
 */
const getPixelData = (context, width, height) => {
  return context.getImageData(0, 0, width, height).data;
};

/**
 * Calculate the height of the text with the given ascent and descent.
 *
 * @param  {object} metrics The object metrics.
 * @param  {number} metrics.descent The position of the descent.
 * @param  {number} metrics.ascent The position of the ascent.
 * @return {number} The height.
 */
const getHeight = ({descent, ascent}) => descent - ascent + 1;

/**
 * Get the ascent of the text.
 *
 * Forward scan the pixel data and get the vertical position of the first non-white pixel.
 *
 * @param  {Uint8ClampedArray} pixelData The RGBA values of the area.
 * @param  {number} width The width of the area.
 * @return {number} The Y coordinate of the ascent.
 */
const getAscent = (pixelData, width) => {
  let index = _.findIndex(pixelData, isNotWhite);

  return getYPosition(index, width);
};

/**
 * Get the descent of the text.
 *
 * Reverse scan the pixel data and get the vertical position of the last non-white pixel.
 *
 * @param  {Uint8ClampedArray} pixelData The RGBA values of the area.
 * @param  {number} width The width of the area.
 * @return {number} The Y coordinate of the descent.
 */
const getDescent = (pixelData, width) => {
  let index = _.findLastIndex(pixelData, isNotWhite);

  return getYPosition(index, width);
};

/**
 * Get the vertical position of the pixel data index, given the width.
 *
 * @param  {number} index The index of the first non-white pixel.
 * @param  {number} width The width of the area.
 * @return {number} The Y coordinate.
 */
const getYPosition = (index, width) => _.floor((index / 4) / width);

/**
 * Check the given RGBA value is below the threshold.
 *
 * @param  {number} value The color channel value (0-255).
 * @return {boolean} True if the value is white.
 */
const isNotWhite = value => value < (255 * THRESHOLD);

/**
 * Convert a physical pixel value to a CSS pixel.
 *
 * @param {number} value The physical pixel.
 * @return {number} The CSS pixel.
 */
const toPx = value => value / DPR;

/**
 * Convert a CSS pixel to a physical pixel.
 *
 * @param {number} value The CSS pixel.
 * @return {number} The physical pixel.
 */
const toDot = value => value * DPR;
